### ISSUE TITLE
`siblingElements()` returns empty collection when build in release mode

### ISSUE THREAD

**Issue by godbout** [scinfu/SwiftSoup#179]
i've tried removing optimization levels for release for SwiftSoup but doesn't help.

the code:
```swift
elements = try document.select(".frontPageWidget tr").first()?.siblingElements()
```

in both debug and release mode the `.first()` element is the same and not nil, but only in debug mode i can grab the siblings. in release mode i get nothing.

just a few months old in Swift/Xcode so it's probably on my side. i've built a library that i use along SwiftSoup and it works fine. but i can't figure out that last one. any idea what am i be doing wrong? thanks!

**Comment by godbout** [comment]
btw using SwiftSoup 2.3.2, Swift 5.3.2, and XCode 12.4.

**Comment by godbout** [comment]
> i've tried removing optimization levels for release for SwiftSoup but doesn't help.

i might believe that i've done that correctly but maybe not. i've made the changes in XCode, but i'm building the release through the command line, with `-c release`, which might be actually enforcing the optimizations. now that i'm thinking about it it kinda makes sense.

**Comment by godbout** [comment]
ok so i've compiled the release through XCode and the issue seems to be with the `Swift Compilter — Code Generation`. if `Optimization Level` is set to something else than `No Optimization`, then `siblingElements()` returns an empty collection. SPM is enforcing the --options which is why i thought it wasn't this at the first place.

**Comment by godbout** [comment]
i was able to create a CLI that works from SPM with following command: `swift build -c release -Xswiftc -Onone`. it passes the `No Optimization` to the Swift Compiler and `siblingElements()` works. the CLI is just 1M bigger than expected, but at least it works :))

**Comment by JaydenIrwin** [comment]
No fix for this yet? This makes my app completely unusable on release builds.

It also effects `.parent()?.html()` which returns empty string on release builds.

**Comment by scinfu** [comment]
@godbout can you provide an HTML to test it?
Thank you

**Comment by godbout** [comment]
hello. (haha.)

sure, here you go: https://gist.github.com/godbout/2fafd288d4c55d36405f93072975755c

works totally fine until i create an archive that turns the optimizations on. at the end i can still compile something, it's just 1 meg higher than needed but ok. but may be nice to find out *what's going on*? (put Marvin Gaye's album on...)

**Comment by scinfu** [comment]
@godbout in this html example what are selecting?
`document.select(".frontPageWidget tr").first()?.siblingElements()`

**Comment by godbout** [comment]
omg what a DUMBASS i am LMAO. made a mistake in the search query. gist updated.

**Comment by scinfu** [comment]
@godbout  work in release mode.
I test in Mac and iOS changing run mode in release and choosing all optimisation levels.
Dhis issue appear only with archive?

**Comment by godbout** [comment]
yeah in my case it fails as soon as it's in RELEASE mode. in DEBUG it finds the items, in RELEASE it doesn't. i'll investigate again though as that's almost a year old issue and i'll post here the results. thanks!

**Comment by godbout** [comment]
tested here, same. works well in DEBUG without optimizations. when i RELEASE it doesn't work, except if i remove the optimizations. will send more info.

**Comment by godbout** [comment]
a video showing the issue:

1. test passes with `No Optimization`, default for `DEBUG`. SwiftSoup returns `siblingElements` properly
2. test fails with `Optimize for Speed`, default for `RELEASE`. SwiftSoup returns properly for `first()` but fails for the `siblingElements()`

https://user-images.githubusercontent.com/121373/149382016-429e02ec-cff6-4fb7-bc28-4c9e0b04d044.mp4

**Comment by godbout** [comment]
the test run in the video is here: https://github.com/godbout/AlfredKat/blob/534c10aaf6c49b48f50c0446777e4508bef3e2b1/Tests/AlfredKatCoreTests/Feature/WorkflowTests.swift#L45
the line failing is here: https://github.com/godbout/AlfredKat/blob/534c10aaf6c49b48f50c0446777e4508bef3e2b1/Sources/AlfredKatCore/Menus/Entrance.swift#L75

maybe it's just me doing something wrong. would not discard this possibility.

**Comment by scinfu** [comment]
Found the bug, parent Node in SwiftSoup is in weak,     [weak var parentNode: Node?](https://github.com/scinfu/SwiftSoup/blob/45a7566f7ec2de0b977f28e825f8df874f210327/Sources/Node.swift#L15)
otherwise we will have a memory leak because parent have an array with all childs.

If you call `document.select(".frontPageWidget tr").first()?.siblingElements()` more time only he last will have zero elements.

document should be put in a variable and set to nil only after use is finished, if it is freed all the elements not retained in memory will be removed.

I'll think of a solution.

**Comment by godbout** [comment]
> Found the bug, parent Node in SwiftSoup is in weak, [weak var parentNode: Node?](https://github.com/scinfu/SwiftSoup/blob/45a7566f7ec2de0b977f28e825f8df874f210327/Sources/Node.swift#L15) otherwise we will have a memory leak because parent have an array with all childs.

makes total sense this is a `weak` var.

> If you call `document.select(".frontPageWidget tr").first()?.siblingElements()` more time only he last will have zero elements.
> 
> document should be put in a variable and set to nil only after use is finished, if it is freed all the elements not retained in memory will be removed.
> 
> I'll think of a solution.

barely dealing with `weak` and `unowned` here so good luck with that one :D

**Comment by scinfu** [comment]
@godbout issue resolved, try now with 2.3.7

**Comment by scinfu** [comment]
Nevermind, i make a rollback with 2.3.8. The issue steel live for now

**Comment by godbout** [comment]
ah, sorry. thought i put this in the email snooze but didn't seem so. my bad.

the fix created other issues?

**Comment by DarkDust** [comment]
I cannot reproduce this with a release build right now, it looks like the Swift compiler has become smarter and is not releasing `document` immediately any more. Even if I split it up into two lines the code works with "Optimise for speed" on my system.

**Comment by godbout** [comment]
nice. will try at my next release.

**Comment by DarkDust** [comment]
> nice. will try at my next release.

If you still run into this issue, please share which Swift compiler version you're using and if possible an example. My guess is that the Swift compiler might make a difference here.

**Comment by godbout** [comment]
> If you still run into this issue, please share which Swift compiler version you're using and if possible an example. My guess is that the Swift compiler might make a difference here.

will do.

**Comment by godbout** [comment]
ok i've tried with the latest version of SwiftSoup and i can't confirm yet whether it works or not, but i can confirm that my binary went from 1.9 MB to 3 MB 😂 me wanting to build in release mode first was to get a smaller binary.

**Comment by godbout** [comment]
FWIW: `swift-driver version: 1.120.5 Apple Swift version 6.1.2 (swiftlang-6.1.2.1.2 clang-1700.0.13.5)
Target: arm64-apple-macosx15.0`

will try later if i can grab all the necessary elements. currently i can't because of external circumstances.

**Comment by aehlke** [comment]
I haven't investigated binary size. A lot of code has been added for optimization work but it's surprising to see a 50% jump. If you know how to instrument that there might be parts we can remove

**Comment by godbout** [comment]
will definitely update you when i work on this. thanks!

**Comment by aehlke** [comment]
Please let me know if you still have this issue and I will reopen and we can organize an investigation. Thank you

**Comment by godbout** [comment]
sorry my bad. it's in my TODO but very low priority. no worries with closing, i'll open back and/or update when i come to test again. thanks!
